src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -79,134 +69,206 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
-            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
+        private static string SdkResolutionExceptionMessage(string methodName) => $"Failed to find all versions of .NET Core MSBuild. Call to {methodName}. There may be more details in stderr.";
+        
+        /// <summary>
+        /// Determines the directory location of the SDK accounting for
+        /// global.json and multi-level lookup policy.
+        /// </summary>
+        private static string? GetSdkFromGlobalSettings(string workingDirectory)
+        {
+            string? resolvedSdk = null;
+            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: DotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
             {
                 if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
-                {
-                    bestSDK = value;
-                }
+                    resolvedSdk = value;
             });
 
-            if (rc == 0 && bestSDK != null)
+            if (rc != 0)
+                throw new InvalidOperationException(SdkResolutionExceptionMessage(nameof(NativeMethods.hostfxr_resolve_sdk2)));
+
+            return resolvedSdk;
+        }
+
+        private static string ResolveDotnetPath()
+        {
+            string? dotnetPath = GetDotnetPathFromROOT();
+
+            if (string.IsNullOrEmpty(dotnetPath))
             {
-                yield return bestSDK;
+                string? dotnetExePath = GetCurrentProcessPath();
+                var isRunFromDotnetExecutable = !string.IsNullOrEmpty(dotnetExePath) 
+                    && Path.GetFileName(dotnetExePath).Equals(ExeName, StringComparison.InvariantCultureIgnoreCase);
+                
+                if (isRunFromDotnetExecutable)
+                    dotnetPath = Path.GetDirectoryName(dotnetExePath);
+                else
+                {
+                    dotnetPath = FindDotnetPathFromEnvVariable("DOTNET_HOST_PATH") 
+                        ?? FindDotnetPathFromEnvVariable("DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR")
+                        ?? GetDotnetPathFromPATH();
+                }
             }
-            else if (rc != 0)
+
+            if (string.IsNullOrEmpty(dotnetPath))
+                throw new InvalidOperationException("Could not find the dotnet executable. Is it set on the DOTNET_ROOT?");
+
+            return dotnetPath;
+        }
+
+        private static string? GetDotnetPathFromROOT()
+        {
+            // 32-bit architecture has (x86) suffix
+            string envVarName = (IntPtr.Size == 4) ? "DOTNET_ROOT(x86)" : "DOTNET_ROOT";
+            var dotnetPath = FindDotnetPathFromEnvVariable(envVarName);
+            
+            return dotnetPath;
+        }
+
+        private static string? GetCurrentProcessPath()
+        {
+            string? processPath = null;
+#if NET6_0_OR_GREATER
```

Comment:
nit: Isn't this always true? Is there a reason for compiling this file against .NET Framework?

Reply:
I thought that #231 retargeted the project to net6.0. Is it really possible for this code to run on net5.0 or lower?

Reply:
Oh, yes. Fixed